File: packages/node-core/src/sql-diagnostics.ts
const DEFAULT_SQL_DIAGNOSTIC_MAX_CHARS = 512;
const SENSITIVE_NAME_RE = /(?:password|passwd|pwd|secret|token|api[_-]?key|access[_-]?key|private[_-]?key|credential|authorization|bearer)/i;

function truncateForDiagnostic(value: string, maxChars = DEFAULT_SQL_DIAGNOSTIC_MAX_CHARS): string {
  if (value.length <= maxChars) return value;
  return `${value.slice(0, maxChars)}…[truncated ${value.length - maxChars} chars]`;
}

function redactSqlLiterals(sql: string): string {
  let result = "";
  let i = 0;
  while (i < sql.length) {
    const ch = sql[i];
    const next = sql[i + 1];
    if (ch === "'" || ch === '"' || ch === "`") {
      const quote = ch;
      result += `${quote}[REDACTED]${quote}`;
      i += 1;
      while (i < sql.length) {
        const current = sql[i];
        if (current === quote) {
          if (sql[i + 1] === quote) {
            i += 2;
            continue;
          }
          i += 1;
          break;
        }
        if (current === "\\" && quote !== "`") {
          i += 2;
        } else {
          i += 1;
        }
      }
      continue;
    }
    if (ch === "$") {
      let j = i + 1;
      while (j < sql.length && sql[j] !== "$") j += 1;
      if (j >= sql.length) {
        result += "$";
        i += 1;
        continue;
      }
      const tag = sql.slice(i, j + 1);
      i = j + 1;
      while (i < sql.length) {
        if (sql.slice(i, i + tag.length) === tag) {
          i += tag.length;
          break;
        }
        i += 1;
      }
      result += "$[REDACTED]$";
      continue;
    }
    if (ch === "-" && next === "-") {
      result += "--[REDACTED_COMMENT]";
      i += 2;
      while (i < sql.length && sql[i] !== "\n" && sql[i] !== "\r") i += 1;
      continue;
    }
    if (ch === "/" && next === "*") {
      result += "/*[REDACTED_COMMENT]*/";
      i += 2;
      while (i < sql.length && !(sql[i] === "*" && sql[i + 1] === "/")) i += 1;
      if (i < sql.length) i += 2;
      continue;
    }
    result += ch;
    i += 1;
  }
  return result;
}

export function redactSqlForDiagnostics(sql: string, maxChars = DEFAULT_SQL_DIAGNOSTIC_MAX_CHARS): string {
  const literalRedacted = redactSqlLiterals(sql);
  const sensitiveRedacted = literalRedacted.replace(/\b([A-Za-z_][\w.-]*)(\s*[:=]\s*)([^\s,;)]+)/g, (match, key: string, separator: string) => {
    if (!SENSITIVE_NAME_RE.test(key)) return match;
    return `${key}${separator}[REDACTED]`;
  });
  return truncateForDiagnostic(sensitiveRedacted, maxChars);
}

export function sqlDiagnosticsEnabled(env: NodeJS.ProcessEnv = process.env): boolean {
  const value = env.DBX_SQL_DEBUG ?? env.DBX_DEBUG_SQL ?? env.DBX_MCP_DEBUG_SQL;
  return value === "1" || value?.toLowerCase() === "true";
}

export function logSqlDiagnostic(scope: string, sql: string, details: Record<string, unknown> = {}, env?: NodeJS.ProcessEnv): void {
  if (!sqlDiagnosticsEnabled(env)) return;
  console.error(`[${scope}] sql:`, JSON.stringify({ ...details, sql: redactSqlForDiagnostics(sql) }));
}

File: crates/dbx-core/src/sql_diagnostics.rs
const DEFAULT_SQL_DIAGNOSTIC_MAX_CHARS: usize = 512;

fn is_sensitive_key(key: &str) -> bool {
    let key = key.to_ascii_lowercase();
    key.contains("password")
        || key.contains("passwd")
        || key == "pwd"
        || key.contains("secret")
        || key.contains("token")
        || key.contains("api_key")
        || key.contains("apikey")
        || key.contains("access_key")
        || key.contains("private_key")
        || key.contains("credential")
        || key.contains("authorization")
        || key.contains("bearer")
}

fn truncate_for_diagnostics(value: String, max_chars: usize) -> String {
    if value.chars().count() <= max_chars {
        return value;
    }
    let head: String = value.chars().take(max_chars).collect();
    let omitted = value.chars().count().saturating_sub(max_chars);
    format!("{head}…[truncated {omitted} chars]")
}

fn redact_literals(sql: &str) -> String {
    let chars: Vec<char> = sql.chars().collect();
    let mut out = String::new();
    let mut i = 0;
    while i < chars.len() {
        let ch = chars[i];
        let next = chars.get(i + 1).copied();
        if matches!(ch, '\'' | '"' | '`') {
            out.push(ch);
            out.push_str("[REDACTED]");
            out.push(ch);
            i += 1;
            while i < chars.len() {
                let current = chars[i];
                if current == ch {
                    if chars.get(i + 1).copied() == Some(ch) {
                        i += 2;
                        continue;
                    }
                    i += 1;
                    break;
                }
                if current == '\\' && ch != '`' {
                    i += 2;
                } else {
                    i += 1;
                }
            }
            continue;
        }
        if ch == '-' && next == Some('-') {
            out.push_str("--[REDACTED_COMMENT]");
            i += 2;
            while i < chars.len() && chars[i] != '\n' && chars[i] != '\r' {
                i += 1;
            }
            continue;
        }
        if ch == '/' && next == Some('*') {
            out.push_str("/*[REDACTED_COMMENT]*/");
            i += 2;
            while i + 1 < chars.len() && !(chars[i] == '*' && chars[i + 1] == '/') {
                i += 1;
            }
            if i + 1 < chars.len() {
                i += 2;
            }
            continue;
        }
        out.push(ch);
        i += 1;
    }
    out
}

fn redact_sensitive_assignments(sql: &str) -> String {
    sql.split_whitespace()
        .map(|part| {
            for separator in ['=', ':'] {
                if let Some((key, _value)) = part.split_once(separator) {
                    if is_sensitive_key(key.trim_matches(|c: char| !c.is_ascii_alphanumeric() && c != '_' && c != '-'))
                    {
                        return format!("{key}{separator}[REDACTED]");
                    }
                }
            }
            part.to_string()
        })
        .collect::<Vec<_>>()
        .join(" ")
}

pub fn redact_sql_for_diagnostics(sql: &str) -> String {
    truncate_for_diagnostics(redact_sensitive_assignments(&redact_literals(sql)), DEFAULT_SQL_DIAGNOSTIC_MAX_CHARS)
}

pub fn debug_sql(scope: &str, sql: &str) {
    log::debug!("[{scope}] sql={}", redact_sql_for_diagnostics(sql));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn redacts_sensitive_literals_and_bounds_large_sql() {
        let sql = format!("select * from users where password = 'secret-123' and api_key=abc {};", "x".repeat(900));
        let redacted = redact_sql_for_diagnostics(&sql);
        assert!(!redacted.contains("secret-123"));
        assert!(!redacted.contains("api_key=abc"));
        assert!(redacted.contains("'[REDACTED]'"));
        assert!(redacted.contains("api_key=[REDACTED]"));
        assert!(redacted.contains("truncated"));
        assert!(redacted.len() < sql.len());
    }
}
